_episodes/05-preprocessor.md

```diff
@@ -0,0 +1,287 @@
+---
+title: "Working with preprocessors"
+teaching: 20?
+exercises: 15?
+questions:
+- "How do I set up a preprocessor?"
+- "Can I use different preprocessors for different variables?"
+- "Can I use different datasets for different variables?"
+objectives:
+- "Create a recipe with multiple preprocessors"
+- "Use different preprocessors for different variables"
+- "Run a recipe with variables from different datasets"
+keypoints:
+- "A recipe can run different preprocessors at the same time."
+- "The setting additional_datasets can be used to add a different dataset."
+- "Variable groups are useful for defining different settings for different variables."
+---
+
+## Preprocessors: What are they and how do they work?
+
+Preprocessing is the process of performing a set of modular operations on the data before applying diagnostics or metrics. In the figure below you see the preprocessor functions in the light blue box on the right.
+
+![figure showing ESMValTool architecture]({{ page.root }}/fig/esmvaltool_architecture.png)
+
+Underneath the hood, each preprocessor is a modular python function that receives an iris cube and sometimes some arguments. The preprocessor applies some mathematical or computational transformation to the input iris cube, then returns the processed iris cube.
+
+## Inspect the example preprocessor
+
+Each preprocessor section includes a preprocessor name, a list of preprocessor steps to be executed and any arguments needed by the preprocessor steps.
+
+>~~~YAML
+> preprocessors:
+>   prep_timeseries:
+>     annual_statistics:
+>       operator: mean
+>~~~
+{: .source}
+
+For instance, the 'annual_statistics' with the  'operation: mean' argument preprocessor receives an iris cube, takes the annual average for each year of data in the cube, and returns the processed cube. 
+
+You could use several preprocessor steps listed in the [documentation](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/recipe/preprocessor.html). The standardised interface between the preprocessors allows them to be used modularly - like lego blocks. Almost any conceivable preprocessing order of operations can be performed using ESMValTool preprocessors. 
```

Comment:
What do you mean with "any conceivable preprocessing order of operations"? I thought there is an exact order how the different tasks in the preprocessor are handled?

Reply:
Never mind. I found the the "custom_order" command just now in your tutorial.  :smile:

Reply:
There is an order in which preprocessors are handled but "any conceivable" here implies "any order you can think of" and that is true if you use custom_order. I found two locations in the Readme docs where there is information for the default order and it is a bit odd that they list the preprocessors quite differently. The first is what showed up with a Google search. This could have implications - for example, the need to regrid before multimodel means is there if we assume the first list of ordering whereas it doesn't need to be for the second list. We have provided the link to the second page but I believe we have to regrid before doing multimodel means, so which is correct? And should we have two locations in the readme docs for this showing different orders? I am calling the attendtion of @hb326 and @LisaBock ('s) to this.
https://docs.esmvaltool.org/projects/ESMValCore/en/latest/recipe/preprocessor.html
https://docs.esmvaltool.org/projects/ESMValCore/en/latest/api/esmvalcore.preprocessor.html

Reply:
I think the second one is the correct order we should link to. In the first link it says that it is "roughly following the default order". so maybe not totally correct as it is sorted in groups.
And as I read, also in the second list `multi_model_statistics` is listed after `regrid`. It is not clear to me what you mean...

Reply:
Ok..I see. I think I mistook regrid_time when reading...so this is ok then?